testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/client/proxy/EncodedParamsProxyTest.java

```diff
@@ -0,0 +1,114 @@
+import javax.ws.rs.client.ClientBuilder;
+import org.jboss.resteasy.client.jaxrs.ResteasyWebTarget;
+import org.jboss.resteasy.test.client.proxy.resource.EncodedParamsProxyInterface;
+import org.jboss.resteasy.test.client.proxy.resource.EncodedParamsProxyResource;
+import org.jboss.resteasy.utils.PortProviderUtil;
+import org.jboss.resteasy.utils.TestUtil;
+import org.jboss.shrinkwrap.api.Archive;
+import org.jboss.shrinkwrap.api.spec.WebArchive;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+/**
+ * @tpSubChapter Resteasy-client
+ * @tpChapter Integration tests
+ * @tpTestCaseDetails Regression test for RESTEASY-1475 and RESTEASY-2068.
+ * @tpSince RESTEasy 3.1.4
+ */
+@RunWith(Arquillian.class)
+@RunAsClient
+public class EncodedParamsProxyTest {
+   private static ResteasyClient client;
+
+   @BeforeClass
+   public static void before() throws Exception {
+      client = (ResteasyClient)ClientBuilder.newClient();
+   }
+
+   @AfterClass
+   public static void after() throws Exception {
+      client.close();
+   }
+
+   @Deployment
+   public static Archive<?> deployUriInfoSimpleResource() {
+      WebArchive war = TestUtil.prepareArchive(EncodedParamsProxyTest.class.getSimpleName());
+      war.addClasses(EncodedParamsProxyInterface.class);
+      return TestUtil.finishContainerPrepare(war, null, EncodedParamsProxyResource.class);
+   }
+
+   private static String generateBaseUrl() {
+      return PortProviderUtil.generateBaseUrl(EncodedParamsProxyTest.class.getSimpleName());
+   }
+
+   /**
+    * @tpTestDetails Verify "/" in "t;hawkular/f;jk-feed?query=100%-%25ig" is sent encoded
+    * @tpSince RESTEasy 3.1.4
+    */
+   @Test
+   public void testEncodeProxy() throws Exception
+   {
+      ResteasyWebTarget target = client.target(generateBaseUrl());
+      EncodedParamsProxyInterface proxy = target.proxy(EncodedParamsProxyInterface.class);
+      Response response = proxy.encode("t;hawkular/f%3Bjk-feed", "100%-%25ig", null);
+      Assert.assertEquals(200, response.getStatus());
+      String uri = response.readEntity(String.class);
+      Assert.assertEquals(generateBaseUrl() + "/test/encode/t;hawkular%2Ff%3Bjk-feed?query=100%25-%25ig", uri);
+   }
+
+   /**
+    * @tpTestDetails Verify "/" in "t;hawkular/f;jk-feed?100-%ig ?" is sent unencoded
+    * @tpSince RESTEasy 3.1.4
+    */
+   @Test
+   public void testNoencodeProxy() throws Exception
+   {
+      ResteasyWebTarget target = client.target(generateBaseUrl());
+      EncodedParamsProxyInterface proxy = target.proxy(EncodedParamsProxyInterface.class);
+      Response response = proxy.noencode("t;hawkular/f%3Bjk-feed", "100%-%25ig", null);
+      Assert.assertEquals(200, response.getStatus());
+      String uri = response.readEntity(String.class);
+      Assert.assertEquals(generateBaseUrl() + "/test/noencode/t;hawkular/f%253Bjk-feed?query=100%25-%2525ig", uri);
```

Comment:
Hi @lyind,

I don't understand this. '/' is not encoded, but '%', when not part of an encoded character, is.

-Ron

Reply:
@ronsigal The method `noencode()` tests the parameter value handling without `@Encoded` annotation. All the tests in `EncodedParamsProxyTest` focus on the proxy client which is why they return the whole URI instead of the parameter values processed by the server.

## Regarding '/'
A `/` in the path parameter value isn't encoded in the original design, which I see as a deliberate decision complementing allowing pattern matching for parameters instead of forcing a fixed 1:1 mapping between path segments and parameters. This was probably done for flexibility reasons.

Examples:
1. ```
   @Path("/{param1:.?}")
   collapsesSlashes(@PathParam("param1") String param1)
   // call via proxy: collapsesSlashes("a/b/c") -> param1 == "a/b/c"
   ```
2. ```
   @Path("/{param1}/{param2}")
   honoursSlashes(@PathParam("param1") String param1, @PathParam("param2") String param2)
   // call via proxy: honoursSlashes("a", "b") -> param1 == "a" && param2 == "b"
   // BUT:
   // call via proxy: honoursSlashes("a/b/c", "d") -> param1 == "a" && param2 == "b"
   ```
3. ```
   @Path("/{param1}/{param2}")
   honoursSlashesEncoded(@Encoded @PathParam("param1") String param1, @PathParam("param2") String param2)
   // call: honoursSlashesEncoded("a/b/c", "d") -> param1 == "a%2Fb%2Fc" && param2 == "b"
   ```

To still be able to use a service with methods looking like in example #3 with *an unknown number of slashes* in parameter values, I made sure the proxy client encodes the slashes to get them through in one piece if it encounters an `@Encoded` `@PathParam`.

To clarify, `noencode()` correctly tests behavior as in example #1.

## Regarding '%'
The `%` must always be encoded according to https://www.ietf.org/rfc/rfc1738.txt :
 1. "The character "%" is unsafe because it is used for encodings of other characters."
 2. "Octets must be encoded if they have no corresponding graphic character within the US-ASCII coded character set, if the use of the corresponding character is unsafe, or if the corresponding character is reserved for some other interpretation within the particular URL scheme."

Since `@Encoded` is not present on `noencode()`'s' `@PathParam` or `@QueryParam` the proxy client may safely assume the values passed are not pre-encoded and just encode the whole string without preserving encoded chars.

Reply:
@ronsigal I just saw your extended explanation at [RESTEASY-2068](https://issues.jboss.org/browse/RESTEASY-2068?focusedCommentId=13699650&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-13699650).

I think confusion is caused by the special handling of `/` in the encoding methods.

Values passed as arguments for `@Encoded` annotated parameters should be encoded already and thus not contain the unencoded slash at all. The example I provided above also missed that fact (I edited it).

I will modify the code to be in line with the handling for non `@Encoded` annotated methods (regarding the slash).

Reply:
@ronsigal Detailed discussion updated in Issue Tracker: [RESTEASY-2068](https://issues.jboss.org/browse/RESTEASY-2068?focusedCommentId=13703720&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-13703720)

Reply:
@ronsigal Can you please have a look at the summary/suggestions I put together in the Jira ticket?